tracer/src/Datadog.Trace/Logging/Internal/Sinks/AsyncConsoleSink.cs

```diff
@@ -0,0 +1,117 @@
+{
+    // in-memory buffer used only from the background thread
+    private readonly StringBuilder _buffer;
+    private readonly StringWriter _bufferWriter;
+
+    private readonly ITextFormatter _textFormatter;
+    private readonly BlockingCollection<LogEvent> _writeQueue;
+    private readonly TextWriter _consoleWriter;
+    private readonly Task _writeTask;
+
+    public AsyncConsoleSink(ITextFormatter formatter, TextWriter consoleWriter, int queueLimit)
+    {
+        // in-memory buffer used only from the background thread
+        _buffer = new StringBuilder(capacity: 512);
+        _bufferWriter = new StringWriter(_buffer);
+
+        _textFormatter = formatter;
+        _writeQueue = new BlockingCollection<LogEvent>(queueLimit);
+        _consoleWriter = consoleWriter;
+
+        _writeTask = Task.Factory.StartNew(
+            WriteToConsoleStream,
+            CancellationToken.None,
+            TaskCreationOptions.LongRunning | TaskCreationOptions.DenyChildAttach,
+            TaskScheduler.Default);
+    }
+
+    public void Emit(LogEvent? logEvent)
+    {
+        // TODO: metrics about dropped logs?
+
+        if (logEvent is null || _writeQueue.IsAddingCompleted || _writeTask.IsCompleted)
+        {
+            // skip null log events.
+            // if we are no longer writing to the console, don't accept any more log events.
+            return;
+        }
+
+        try
+        {
+            // try to add the log event to the queue (non-blocking).
+            // if the queue is full, the log event is dropped.
+            _writeQueue.TryAdd(logEvent);
+        }
+        catch (InvalidOperationException)
+        {
+            // queue is marked as complete, drop log event.
+            // this won't usually happen as we check IsAddingCompleted above,
+            // but it can theoretically happen if the queue is completed from another thread
+            // after we checked IsAddingCompleted but before we called TryAdd().
+        }
+    }
+
+    private void WriteToConsoleStream()
+    {
+        try
+        {
+            // GetConsumingEnumerable() blocks until an item
+            // is available or the collection is marked as complete.
+            // The foreach loop will exit when the collection is marked as complete.
+            foreach (var logEvent in _writeQueue.GetConsumingEnumerable())
+            {
+                // clear in-memory buffer and format event into buffer
+                _buffer.Clear();
+                _textFormatter.Format(logEvent, _bufferWriter);
+
+                // In more recent runtimes, this uses the new WriteLine(StringBuilder) overload which
+                // uses StringBuilder.GetChunks() without allocating a new string.
+                // In earlier runtimes, the WriteLine(object) overload calls StringBuilder.ToString() instead.
+                _consoleWriter.Write(_buffer);
+                _consoleWriter.Flush();
```

Comment:
I think this is not required for a console stdout textwriter.

https://source.dot.net/#System.Console/System/ConsolePal.Unix.ConsoleStream.cs,52

https://github.com/microsoft/referencesource/blob/f7df9e2399ecd273e90908ac11caf1433e142448/mscorlib/system/io/__consolestream.cs#L109